### PR TITLE
try creating a system test and running in Travis

### LINKED ISSUE #3683: system tests (rails 5.1+ feature)
New testing framework in rails allows tests to take screenshots! But also cool form filling syntax:

https://chriskottom.com/blog/2017/04/full-stack-testing-with-rails-system-tests/

### PR DESCRIPTION
**Pull request by jywarren**
Closes #3683

https://guides.rubyonrails.org/testing.html#implementing-a-system-test

https://chriskottom.com/blog/2017/04/full-stack-testing-with-rails-system-tests/

### CODE REVIEW

**Comment by jywarren** [comment]
Needs chromedriver! https://dev.to/aergonaut/running-rails-5-system-tests-on-travis-ci-with-chromedriver-4nm7

**Comment by jywarren** [comment]
https://stackoverflow.com/questions/20364674/chromedriver-on-travis-ci

**Comment by jywarren** [comment]
https://stackoverflow.com/questions/42210295/travis-unable-to-find-chromedriver?rq=1

**Comment by jywarren** [comment]
Selenium::WebDriver::Error::WebDriverError:         Selenium::WebDriver::Error::WebDriverError:  Unable to find chromedriver. Please download the server from http://chromedriver.storage.googleapis.com/index.html and place it somewhere on your PATH. More info at https://github.com/SeleniumHQ/selenium/wiki/ChromeDriver.

**Comment by jywarren** [comment]
Hmm, @siaw23 have you ever done this? I think this could dramatically stabilize our testing of front-end features!

**Comment by siaw23** [comment]
I took a shot at this. System tests have been fixed but only locally(on development). I think we can't run system tests on Travis if we're using Passenger because Passenger runs multiple process which won't work with Capybara. So we can either change from Puma to Passenger on production (which I recommend) or run system tests only on development with Puma 😬...

**Comment by jywarren** [comment]
Well, system tests should run in test mode, not development mode, right? Sorry, I'm new to this! Right now tests don't run on passenger at all, they are lower level. And we don't run them in production mode at all. So don't we not lose anything in running system tests on Puma?

Thanks!!!

**Comment by jywarren** [comment]
@publiclab/plots2-reviewers anyone want to take a crack at getting these to run in Travis? The above posts seem to show it's possible.

**Comment by siaw23** [comment]
Oh I'll have to check this again! I left this like this thinking the usual tests (non system tests) are running.

**Comment by siaw23** [comment]
OK so I had to re-read what you wrote:

>  Right now tests don't run on passenger at all, they are lower level.

They do. `884/884: [===============================] 100% Time: 00:03:01, Time: 00:03:01`


>  And we don't run them in production mode at all.

We ran the normal tests like usual. They only test we don't run now are system tests and that's because of Passenger. So in this PR we're running normal tests + system tests in development and test modes. Then on production, we're just running the normal tests. Though ideally we'd like to run system tests on Travis too.


> So don't we not lose anything in running system tests on Puma?

I'm not sure what this means but I don't see how we'd lose anything on running system tests on puma, it's only to test UI right?

File: .travis.yml [hunk 1]
```diff
@@ -30,7 +30,9 @@ install:
 script:
   - docker-compose exec web bash -c "CI=TRUE GENERATE_REPORT=true rake test:all"
   - docker-compose exec web bash -c "CI=TRUE GENERATE_REPORT=true rails test -d"
-  - docker-compose exec web bash -c "CI=TRUE GENERATE_REPORT=true rails test:system"
+  # Capybara can't work with Passenger on Travis, hence disabling system test on
+  # production until we change over to a Rack server.
+  # - docker-compose exec web bash -c "CI=TRUE GENERATE_REPORT=true rails test:system"
```
Comment (jywarren) [inline .travis.yml]: So I think we really need to re-enable this line, and to follow the guidance in the links I provided to try to get system tests to run in Travis.

File: README.md [hunk 2]
```diff
@@ -93,7 +93,7 @@ For information on how to install for use with the cloud environment, please see
 9. By default, start rails with `passenger start` from the Rails root and open http://localhost:3000 in a web browser.
 (for local SSL work, see [SSL](#ssl+in+development) below)
 10. Wheeeee! You're up and running! Log in with test usernames "user", "moderator", or "admin", and password "password".
-11. Run `rails test -d` to confirm that your install is working properly.
+11. Run `rails test` to confirm that your install is working properly. Or `rails test:system` for system tests.
```
Comment (jywarren) [inline README.md]: what was the `-d` originally?

**Comment by jywarren** [comment]
Anyways, my biggest interest in this PR is to get system tests running in Travis, so we can use them as part of our CI. Where would we set it to use Puma? 

https://dev.to/aergonaut/running-rails-5-system-tests-on-travis-ci-with-chromedriver-4nm7 seems to show it working in Travis. 

Actually the errors I had seen were about not being able to find ChromeDriver, not related to Capybara. But maybe you got past those errors and saw a Capybara/Passenger error? If so, could you point me at the commit where you saw that?

File: README.md (hunk 2 above)
Comment (siaw23) [reply README.md]: `-d` = `Output test failures and errors after the test run`

**Comment by sashadev-sky** [comment]
@jywarren @siaw23 I am looking through this right now. To try to run this I can just fetch your PR locally and then push to this branch?

**Comment by siaw23** [comment]
Yes you can @sashadev-sky

**Comment by sashadev-sky** [comment]
@jywarren @siaw23 Ok I am new to a bunch of the things mentioned here so my PRs are probably just going to be trial and error at first so feel free to ignore them

**Comment by sashadev-sky** [comment]
@jywarren @siaw23 I changed some setup configurations and I have it working locally fine now. Is the issue just getting them into Travis? I think we can just make a rake task for that?

**Comment by jywarren** [comment]
Yes, i think that's right - and in Travis it has to be using the configured chromedriver and chrome from either `apt` or the premade Travis versions. The examples near the top (blog posts and such) seem to show that it has worked at least for some people, although perhaps we're missing something about Capybara/Puma or something in the setup as @siaw23 has pointed out. 

My guess is that if there's a way we can configure Travis to run them either as they are or with Puma (if that's the missing part) then that would be best. 

This will hopefully be really great for JS testing. The hidden treat is also that it may be possible to get it to generate a set of screenshots of different pages automatically 💥

**Comment by siaw23** [comment]
@jywarren I have seen those posts you referred to. I **KNOW** system tests are possible on Travis. What the articles you referred to don't mention is what server they are using. They only showed how it works on Travis without making mention of which server they are using. With Thin, Webrick, Puma that article is relevant, with Passenger (which is not a Rack server...I think, it doesn't work).

**Comment by jywarren** [comment]
Hi @siaw23 -- so sorry, i know you know! Sincere apologies, didn't mean to run through the same info for you, just trying to catch Sasha up a bit. Thank you, that's an excellent point on their omission of the server in those posts. 

What your observation makes me wonder is whether we can override the Passenger server config with a Rack server solely in the Travis context, for the purposes of the Travis CI. Do you think that'd be a reasonable way to achieve system tests in Travis, or do you think we should consider another path?  Thank you! ❤️

**Comment by sashadev-sky** [comment]
@jywarren yes I have been playing with the screenshots, it's really great. teaspoon mocha sinon with rails fixtures had my head spinning so i'm excited about getting this to work. 

https://www.rubydoc.info/github/jnicklas/capybara#The_DSL this page is the capybara docs and it has every possible way you can configure it and pros and cons, etc. 

Like for ex., I got it to run with `poltergeist` driver because `poltergeist` relies on `phantomjs`, but the issue with that is it gets weird about file loading order and could also have discrepancies with Travis. And then you need to go through and see all of the configuration options its offering to try to fix that. (And also `phantomjs` is not maintained by anyone anymore)

So instead I tried `RSpec` and got it running really easily, but I realized I think the results might only work with static data (or I am not writing my tests correctly, TBD) and they have their own recommendations about configurations and what drivers to add that we would need to go through...

So yeah I agree out of this list of options there could definitely be a solution that doesn't involve infrastructure changes but we would need to dig a bit.

**Comment by siaw23** [comment]
I have no experience with running system tests + Passenger + Travis. I saw a comment on Passenger or Travis' repo once stating how this may not be possible. Couldn't find it unfortunately when I tried to search for it. If we used something like Puma we could avoid a lot of headaches for contributors.

So for this particular issue and with my commits, we can run system tests locally until we switch to a Rack server to have them run on Travis. It'll be ideal to have system test running on CI. I honestly don't know how to do it with Passenger + Travis.

**Comment by sashadev-sky** [comment]
@siaw23 Yeah sorry me too I didn't see your comment until after I had posted my really long one 🙈I will be so happy the day travis CI makes perfect sense to me

**Comment by jywarren** [comment]
Hi, all - in case it helps us get this running in Travis, here's one environment where this has worked; definitely using Puma: https://travis-ci.org/JunichiIto/rails-5-1-system-test-sandbox/builds/223785193#L193

**Comment by jywarren** [comment]
Trying to reproduce the error I'd seen where Travis couldn't find chomedriver, and want to try to get past that  -- we can also add this to get past it, i hope:

```
script:
  - whereis google-chrome-stable
  - whereis chromedriver
```

Looking at https://www.amihaiemil.com/2017/07/14/selenium-headless-chrome-travis.html and https://github.com/opencharles/charles/blob/master/.travis.yml

If we get past it, i'd like to consider running tests on Puma to get past the Passenger issue, but let's see how this does first.

**Comment by jywarren** [comment]
```
Warning: You're using an unsupported version of selenium-webdriver, please upgrade.
Warning: You're using an unsupported version of selenium-webdriver, please upgrade.
ERROR["test_searching_an_item_from_the_homepage", #<Minitest::Reporters::Suite:0x00005626b09a8828 @name="SearchTest">, 0.3501734880001095]
 test_searching_an_item_from_the_homepage#SearchTest (0.35s)
Selenium::WebDriver::Error::WebDriverError:         Selenium::WebDriver::Error::WebDriverError:  Unable to find chromedriver. Please download the server from http://chromedriver.storage.googleapis.com/index.html and place it somewhere on your PATH. More info at https://github.com/SeleniumHQ/selenium/wiki/ChromeDriver.
```

**Comment by SamueleEsposito** [comment]
@jywarren Correct me if I'm wrong, but I think the test are failing because they are being run in a Docker container, while chromedriver is installed on the host system.

**Comment by SamueleEsposito** [comment]
@jywarren I could be wrong, but I think you are installing chromedriver on the host system, while the tests are being run in a docker container.

**Comment by jywarren** [comment]
```
whereis google-chrome-stable
google-chrome-stable: /usr/bin/google-chrome-stable /usr/bin/X11/google-chrome-stable
The command "whereis google-chrome-stable" exited with 0.
0.01s$ whereis chromedriver
chromedriver:
The command "whereis chromedriver" exited with 0.
```

**Comment by jywarren** [comment]
Ooh, interesting, thanks for the tip! Should I wrap the chromedriver I stall command in a docker compose statement? I had seen a few different ways to install chromedriver in the above tutorials. Does one look better to you? Thanks!

**Comment by SamueleEsposito** [comment]
> Should I wrap the chromedriver I stall command in a docker compose statement?

Yeah, that should work.

> I had seen a few different ways to install chromedriver in the above tutorials. Does one look better to you?

I'd use apt

**Comment by jywarren** [comment]
OK tried via travis `addon` and `apt` commands, and put the `which` queries into docker-container commands so we can see where it gets installed if at all.

**Comment by SamueleEsposito** [comment]
I don't think this will work @jywarren. You need to run apt install within the docker-compose command

**Comment by jywarren** [comment]
Hmm, i'm not seeing the Travis build even getting created from that last commit, even after retriggering it. I'll try `apt-get` inside the container... maybe we can circle back to that.

**Comment by jywarren** [comment]
installing with apt-get inside docker compose got:

```

1.59s$ docker-compose exec web bash -c "apt-get install chromium-chromedriver"
The SECRET_KEY_BASE variable is not set. Defaulting to a blank string.
Reading package lists... Done
Building dependency tree       
Reading state information... Done
E: Unable to locate package chromium-chromedriver
The command "docker-compose exec web bash -c "apt-get install chromium-chromedriver"" exited with 100.
```

odd... maybe the version of debian we're on isn't up to date enough. I'm reverting to try the travis `apt` approach to see if we can get that to build this time. Then i'll look up the right command for this version of debian.

**Comment by jywarren** [comment]
Also see: https://github.com/travis-ci/travis-ci/issues/272, https://500tech.com/blog/all/running-travis-ci-unit-tests-using-google-chrome

**Comment by jywarren** [comment]
Tested with https://jpmelos.com/articles/how-use-chrome-selenium-inside-docker-container-running-python/ installing using `wget` and `dpkg`...

**Comment by SamueleEsposito** [comment]
@jywarren I think you forgot to unzip the file

**Comment by jywarren** [comment]
lol thanks for catching my really silly mistakes. I'm doing like 30 seconds
of work on this every 2 days... :-)

**Comment by jywarren** [comment]
Needs rebase!

**Comment by jywarren** [comment]
Whoaaa....

```

$ docker-compose exec web bash -c "curl https://chromedriver.storage.googleapis.com/2.31/chromedriver_linux64.zip"
The SECRET_KEY_BASE variable is not set. Defaulting to a blank string.
�<�C�chromedriver��y|TE��o��v�"��hX�
H $a���t��Nw�I�41�����dwE�®��ˈ����a��
�K�sN���nR�3�3��7|�uߪ[U���v�W�eL���pm��jݡD҉�G��-�&QK����k��V�[�t�ZbP���2]��SrgxbP�[���0Fh�'1(�*_jQM�a]}�'޿3(����z.\>�5Jؿ5foPx@��@�}-�u��wH��0F��a��'�[O�ǖ���-�
O�
��f�s�������C�.�OdH�-�S=Wwo��P6�6�fͻv[�0���-V�p��kGw;���z`:��)�gQz��c���L�fx����P���:|�H\�O*�i�<�Q��a��?A�S|���V�
^��._���)�
�C��V�C
������
~V�#�m���S��
>^��+�B�(�FN�w(�~�R�/(x���^
�W�*�X�Rp��[|���S��R�_P��
```

**Comment by jywarren** [comment]
Uhhh i guess curl is outputting the zip?

**Comment by jywarren** [comment]
```
...
ï¿½<ï¿½Cï¿½ï¿½ï¿½chromedriverPK:Iï¿½<travis_time:end:284701c0:start=1553906265833651377,finish=1553906295762015167,duration=29928363790
[0K[32;1mThe command "docker-compose exec web bash -c "curl https://chromedriver.storage.googleapis.com/2.31/chromedriver_linux64.zip"" exited with 0.[0m

travis_time:start:026baf82
[0K$ docker-compose exec web bash -c "unzip chromedriver_linux64.zip"
The SECRET_KEY_BASE variable is not set. Defaulting to a blank string.
unzip:  cannot find or open chromedriver_linux64.zip, chromedriver_linux64.zip.zip or chromedriver_linux64.zip.ZIP.
```

**Comment by jywarren** [comment]
i'll just switch to wget

**Comment by jywarren** [comment]
Yes!!! Chromedriver installed, only:

```
Selenium::WebDriver::Error::WebDriverError:         Selenium::WebDriver::Error::WebDriverError: unable to connect to chromedriver 127.0.0.1:9515
```

Need to config address and port. 

Google chrome didn't install, there was an error:

```
1.76s$ docker-compose exec web bash -c "dpkg -i google-chrome-stable_current_amd64.deb --fix-missing; apt-get -fy install"
The SECRET_KEY_BASE variable is not set. Defaulting to a blank string.
dpkg: error: cannot access archive '--fix-missing': No such file or directory
Reading package lists... Done
Building dependency tree       
Reading state information... Done
0 upgraded, 0 newly installed, 0 to remove and 44 not upgraded.
The command "docker-compose exec web bash -c "dpkg -i google-chrome-stable_current_amd64.deb --fix-missing; apt-get -fy install"" exited with 0.
0.72s$ docker-compose exec web bash -c "whereis google-chrome-stable"
The SECRET_KEY_BASE variable is not set. Defaulting to a blank string.
google-chrome-stable:
```

**Comment by jywarren** [comment]
Wow!! Everything built but there was a config error in the test itself:

```
Warning: You're using an unsupported version of selenium-webdriver, please upgrade.
Warning: You're using an unsupported version of selenium-webdriver, please upgrade.
ERROR["test_searching_an_item_from_the_homepage", #<Minitest::Reporters::Suite:0x000055eb80641088 @name="SearchTest">, 0.8642308769999545]
 test_searching_an_item_from_the_homepage#SearchTest (0.86s)
ArgumentError:         ArgumentError: unknown option: {:binary=>"/usr/bin/chromedriver"}
            test/system/search_test.rb:6:in `block in <class:SearchTest>'
```

File: Gemfile [hunk 3]
```diff
@@ -111,6 +113,7 @@ group :test, :development do
   gem 'rake',  '~> 12.3.2'
   gem 'rest-client'
   gem 'rspec'
+  gem 'selenium-webdriver', '~> 3.3.0'
```
Comment (jywarren) [inline Gemfile]: Should try latest https://rubygems.org/gems/selenium-webdriver/versions

Comment (jywarren) [reply Gemfile]: 3.141.0

**Comment by jywarren** [comment]
https://github.com/SeleniumHQ/selenium/wiki/Ruby-Bindings

**Comment by jywarren** [comment]
Closer....!

```
Warning: You're using an unsupported version of selenium-webdriver, please upgrade.
Warning: You're using an unsupported version of selenium-webdriver, please upgrade.
ERROR["test_searching_an_item_from_the_homepage", #<Minitest::Reporters::Suite:0x0000561b1b6e04b0 @name="SearchTest">, 121.06563903400001]
 test_searching_an_item_from_the_homepage#SearchTest (121.07s)
Selenium::WebDriver::Error::UnknownError:         Selenium::WebDriver::Error::UnknownError: unknown error: Chrome failed to start: exited abnormally
          (Driver info: chromedriver=2.35.528139 (47ead77cb35ad2a9a83248b292151462a66cd881),platform=Linux 4.4.0-101-generic x86_64)
            test/system/search_test.rb:6:in `block in <class:SearchTest>'
```

I'll try a newer version of selenium-webdriver

**Comment by jywarren** [comment]
Nope, although we got to the same point with a more recent version so that's good.

File: .travis.yml [hunk 4]
```diff
@@ -27,9 +28,31 @@ install:
   - 'if [ "$TRAVIS_PULL_REQUEST" != "false" ]; then gem install danger danger-junit; fi'
 
 script:
+  - docker-compose exec web bash -c "wget https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb"
+  - docker-compose exec web bash -c "dpkg -i google-chrome-stable_current_amd64.deb; apt-get -fy install"
+#  - docker-compose exec web bash -c "apt-get install google-chrome-stable chromium-chromedriver"
+  - docker-compose exec web bash -c "whereis google-chrome-stable"
+  - docker-compose exec web bash -c "google-chrome-stable --headless --disable-gpu --remote-debugging-port=9222 http://localhost &"
+  - docker-compose exec web bash -c "wget https://chromedriver.storage.googleapis.com/2.35/chromedriver_linux64.zip"
```
Comment (jywarren) [inline .travis.yml]: Should we switch to v73???

https://sites.google.com/a/chromium.org/chromedriver/


https://chromedriver.storage.googleapis.com/index.html?path=73.0.3683.68/

But there's also troubleshooting info on that first page

Comment (jywarren) [reply .travis.yml]: Try `73.0.3683.68` next

**Comment by jywarren** [comment]
Ooh, a more descriptive error:

```
[/usr/local/bundle/gems/selenium-webdriver-3.141.0/lib/selenium/webdriver/remote/response.rb:32]: Selenium::WebDriver::Error::UnknownError: unknown error: Chrome failed to start: exited abnormally (unknown error: DevToolsActivePort file doesn’t exist) (The process started from chrome location /usr/bin/google-chrome is no longer running, so ChromeDriver is assuming that Chrome has crashed.) (Driver info: chromedriver=73.0.3683.68 (47787ec04b6e38e22703e856e101e840b65afe72),platform=Linux 4.4.0-101-generic x86_64) test/system/search_test.rb:6:in ‘block in '
```

File: test/application_system_test_case.rb [hunk 5]
```diff
@@ -0,0 +1,8 @@
+require "test_helper"
+
+class ApplicationSystemTestCase < ActionDispatch::SystemTestCase
+  caps = Selenium::WebDriver::Remote::Capabilities.chrome("chromeOptions" => {"args" => %w(--headless)})
```
Comment (jywarren) [inline test/application_system_test_case.rb]: Maybe we need to set chrome flags here instead of in .travis.yml?

**Comment by jywarren** [comment]
uhhhhhhhh.......

**Comment by jywarren** [comment]
🎉 🎉 🎉 ✔️ 

```
Capybara starting Puma...
* Version 3.12.1 , codename: Llamas in Pajamas
* Min threads: 0, max threads: 4
* Listening on tcp://127.0.0.1:38566
1 items added from related_and_hyphenated_terms.dict.txt
1 items added from related_and_hyphenated_terms.dict.txt
  1/1: [===================================] 100% Time: 00:00:08, Time: 00:00:08
Finished in 8.53907s
1 tests, 1 assertions, 0 failures, 0 errors, 0 skips
```

**Comment by jywarren** [comment]
Now I guess we should see what we can remove and still let this run... to simplify!

**Comment by jywarren** [comment]
Anything else we can remove?

File: .travis.yml [hunk 6]
```diff
@@ -27,9 +27,18 @@ install:
   - 'if [ "$TRAVIS_PULL_REQUEST" != "false" ]; then gem install danger danger-junit; fi'
 
 script:
+  - docker-compose exec web bash -c "wget https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb"
+  - docker-compose exec web bash -c "dpkg -i google-chrome-stable_current_amd64.deb; apt-get -fy install"
+  - docker-compose exec web bash -c "nohup Xvfb -ac :9 -screen 0 1280x1024x16 &"
+  - docker-compose exec web bash -c "export DISPLAY=:99.0"
```
Comment (jywarren) [inline .travis.yml]: Let's try removing these and also add links for where to update dependencies

File: Gemfile [hunk 7]
```diff
@@ -36,7 +36,6 @@ gem 'omniauth-github', '~> 1.1', '>= 1.1.2'
 gem 'omniauth-google-oauth2'
 gem 'omniauth-twitter'
 gem "paperclip", "~> 6.1.0"
-gem 'passenger'
```
Comment (jywarren) [inline Gemfile]: Oh let's try re adding this?

**Comment by jywarren** [comment]
OK this is pretty awesome. I've refined it down to what I believe is the minimum changes needed and cleaned up comments. This should be ready for a merge and then we can start building:

1. more system tests for critical systems
2. (potentially) a screenshot-generating test :-)

**Comment by jywarren** [comment]
This is ready to merge if it passes tests. Opening next steps here: https://github.com/publiclab/plots2/issues/5316

**Comment by jywarren** [comment]
🎉 🎉 🎉

**Pull request merged**